src/api.rs

```diff
@@ -64,6 +64,9 @@ impl ManifestRequest {
 	}
 
 	fn storage_path(&self, ns: &str) -> String {
+		if ns == self.image.as_ref().split("/").next().unwrap_or_default() {
+			return format!("manifests/{}/{}", self.image, self.reference);
+		}
 		format!("manifests/{}/{}/{}", ns, self.image, self.reference)
```

Comment:
I would use a match here, as well as `splitn` and a char instead of str, something like
```suggestion
		match self.image.as_ref().splitn(2, '/').next() {
			Some(part) if part == ns => format!("manifests/{}/{}", self.image, self.reference),
			_ => format!("manifests/{}/{}/{}", ns, self.image, self.reference
			// Some(_) | None => format!("manifests/{}/{}/{}", ns, self.image, self.reference would also work
		}
```

It likely compiles down to identical bytecode (with the possible exception of using `'/'` instead of `"/"`, but this is more idiomatic/common Rust code